src/Compilers/CSharp/Portable/Compilation/CSharpCompilation.cs

```diff
@@ -3821,6 +3820,279 @@ internal new NamedTypeSymbol CreateNativeIntegerTypeSymbol(bool signed)
+                    this.IsReadOnlySpanType(csharpLeftType) &&
+                    this.IsReadOnlySpanType(csharpRightType))
+                {
+                    return;
+                }
+
+                throw new ArgumentException($"Unsupported built-in binary operator: {csharpReturnType.ToDisplayString()} operator {name}({csharpLeftType.ToDisplayString()}, {csharpRightType.ToDisplayString()})");
+            }
+        }
+
+        protected override IMethodSymbol CommonCreateBuiltinOperator(
+            string name,
+            ITypeSymbol returnType,
+            ITypeSymbol operandType)
+        {
+            var csharpReturnType = returnType.EnsureCSharpSymbolOrNull(nameof(returnType));
+            var csharpOperandType = operandType.EnsureCSharpSymbolOrNull(nameof(operandType));
+
+            // caller already checked all of these were not null.
+            Debug.Assert(csharpReturnType is not null);
+            Debug.Assert(csharpOperandType is not null);
+
+            var syntaxKind = SyntaxFacts.GetOperatorKind(name);
+            if (syntaxKind == SyntaxKind.None || name is WellKnownMemberNames.TrueOperatorName or WellKnownMemberNames.FalseOperatorName)
+                throw new ArgumentException($"Illegal operator name '{name}'", nameof(name));
+
+            var unaryOperatorName = OperatorFacts.UnaryOperatorNameFromSyntaxKindIfAny(syntaxKind, SyntaxFacts.IsCheckedOperator(name));
+            if (unaryOperatorName != name)
+                throw new ArgumentException($"'{name}' was not a valid unary operator name", nameof(name));
+
+            // Lang specific checks to ensure this is an acceptable operator.
+            checkOperatorKinds();
+
+            return new SynthesizedIntrinsicOperatorSymbol(csharpOperandType, name, csharpReturnType).GetPublicSymbol();
+
+            static bool isDynamicOrError(TypeSymbol type)
+                => type.TypeKind is TypeKind.Dynamic or TypeKind.Error;
+
+            void checkOperatorKinds()
+            {
+                // Dynamic built-in operators allow virtually all operations with all types.  So we do no further checking here.
+                if (isDynamicOrError(csharpReturnType) ||
+                    isDynamicOrError(csharpOperandType))
+                {
+                    return;
+                }
+
+                // Use fast-path check to see if this types are ok.
+                var unaryKind = OperatorFacts.SyntaxKindToUnaryOperatorKind(SyntaxFacts.GetPrefixUnaryExpression(syntaxKind));
+                var easyOutUnaryKind = OverloadResolution.UnopEasyOut.OpKind(unaryKind, csharpOperandType);
+
+                if (easyOutUnaryKind != UnaryOperatorKind.Error)
+                {
+                    var signature = this.builtInOperators.GetSignature(easyOutUnaryKind);
+                    if (TypeSymbol.Equals(csharpReturnType, signature.ReturnType, TypeCompareKind.ConsiderEverything) &&
```

Comment:
>TypeSymbol.Equals(csharpReturnType, signature.ReturnType, TypeCompareKind.ConsiderEverything) [](http://example.com/codeflow?start=24&length=93)

Does this prevent nullable types? Do we have a test for a scenario like that?

Reply:
yes. CreateBuiltinBinaryOperator_NotSupported checks for nullable types in any position.